drivers/interrupt_controller/intc_ite_it8xxx2.c

```diff
@@ -69,6 +69,131 @@ inline void set_csr(unsigned long bit)
+	uint8_t div_uart;
+	uint8_t div_smb;
+	uint8_t div_sspi;
+	uint8_t div_ec;
+	uint8_t div_jtag;
+	uint8_t div_pwm;
+	uint8_t div_usbpd;
+};
+
+static const struct pll_config_t pll_configuration[] = {
+	/*
+	 * PLL frequency setting = 4 (48MHz)
+	 * FND   div = 0 (PLL / 1 = 48 mhz)
+	 * UART  div = 1 (PLL / 2 = 24 mhz)
+	 * SMB   div = 1 (PLL / 2 = 24 mhz)
+	 * SSPI  div = 1 (PLL / 2 = 24 mhz)
+	 * EC    div = 6 (FND / 6 =  8 mhz)
+	 * JTAG  div = 1 (PLL / 2 = 24 mhz)
+	 * PWM   div = 0 (PLL / 1 = 48 mhz)
+	 * USBPD div = 5 (PLL / 6 =  8 mhz)
+	 */
+	{.pll_freq  = 4,
+	 .div_fnd   = 0,
+	 .div_uart  = 1,
+	 .div_smb   = 1,
+	 .div_sspi  = 1,
+	 .div_ec    = 6,
+	 .div_jtag  = 1,
+	 .div_pwm   = 0,
+	 .div_usbpd = 5}
+};
+
+void __intc_ram_code chip_pll_ctrl(enum chip_pll_mode mode)
+{
+	IT8XXX2_ECPM_PLLCTRL = mode;
+	/* for deep doze / sleep mode */
+	IT8XXX2_ECPM_PLLCTRL = mode;
+}
+
+void __intc_ram_code chip_run_pll_sequence(const struct pll_config_t *pll)
+{
+	/* Enable HW timer to wakeup chip from the sleep mode */
+	timer_5ms_one_shot();
+	/*
+	 * Configure PLL clock dividers.
+	 * Writing data to these registers doesn't change the
+	 * PLL frequency immediately until the status is changed
+	 * into wakeup from the sleep mode.
+	 * The following code is intended to make the system
+	 * enter sleep mode, and wait HW timer to wakeup chip to
+	 * complete PLL update.
+	 */
+	IT8XXX2_ECPM_PLLFREQR = pll->pll_freq;
+	/* Pre-set FND clock frequency = PLL / 3 */
+	IT8XXX2_ECPM_SCDCR0 = (2 << 4);
+	/* JTAG and EC */
+	IT8XXX2_ECPM_SCDCR3 = (pll->div_jtag << 4) | pll->div_ec;
+	/* Chip sleep after wait for interrupt (wfi) instruction */
+	chip_pll_ctrl(CHIP_PLL_SLEEP);
+	/* Chip sleep and wait timer wake it up */
+	__asm__ volatile  ("wfi");
+	/* New FND clock frequency */
+	IT8XXX2_ECPM_SCDCR0 = pll->div_fnd << 4;
+	/* Chip doze after wfi instruction */
+	chip_pll_ctrl(CHIP_PLL_DOZE);
+	/* UART */
+	IT8XXX2_ECPM_SCDCR1 = pll->div_uart;
+	/* SSPI and SMB */
+	IT8XXX2_ECPM_SCDCR2 = (pll->div_sspi << 4) | pll->div_smb;
+	/* USBPD and PWM */
+	IT8XXX2_ECPM_SCDCR4 = (pll->div_usbpd << 4) | pll->div_pwm;
+}
+
+static void chip_configure_pll(const struct pll_config_t *pll)
+{
+	/* Re-configure PLL clock or not. */
+	if (((IT8XXX2_ECPM_PLLFREQR & 0xf) != pll->pll_freq) ||
+		((IT8XXX2_ECPM_SCDCR0 & 0xf0) != (pll->div_fnd << 4)) ||
+		((IT8XXX2_ECPM_SCDCR3 & 0xf) != pll->div_ec)) {
+#ifdef CONFIG_ESPI
+		/*
+		 * TODO: implement me
+		 * We have to disable eSPI pad before changing
+		 * PLL sequence or sequence will fail if CS# pin is low.
+		 */
+#endif
+		/* Run change PLL sequence */
+		chip_run_pll_sequence(pll);
+#ifdef CONFIG_ESPI
+		/*
+		 * TODO: implement me
+		 * Enable eSPI pad after changing PLL sequence
+		 */
+#endif
+	}
+}
+
+static int chip_change_pll(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* store enabled interrupts of chip-level */
+	for (int i = 0; i < ARRAY_SIZE(reg_enable); i++)
+		ier_setting[i] = *reg_enable[i];
```

Comment:
Can you use irq_lock()/irq_unlock() here instead?  That would make this code independent of the INTC driver.

Reply:
I can't disable global interrupt here, because we need one-shot timer's interrupt (will be enabled later) to wake up chip from sleep mode to complete changing PLL sequence.

We disable all chip-level interrupts here is because we don't want to a pending interrupt wake up chip early to break the sequence.